### PR TITLE
Add support for function contracts in cprover_bindings.

### PR DESCRIPTION
**Pull request by monal**
### Description of changes: 

Implement `CodeWithContract` type in `cprover_bindings` to generate Ireps of type `Code` but with additional fields for  handling function contracts. Adds the `#spec_requires` and `#spec_ensures` clauses to the Irep.

### Call-outs:

Added a new type `Spec` with field `clauses:Vec<Expr>` to store a list of boolean expressions that would serve as pre (or post) conditions on the function. This allows us to handle required `#[PartialEq]` trait for the clauses. 

### Testing:

* How is this change tested?
By running kani regression tests. As I haven't added the `#[kani::requires]` and `#[kani::ensures]` macros on the Rust side yet, it is not straightforward to add additional test cases right now.

* Is this a refactor change?
No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by celinval** [comment]
Can you push this to a feature branch for now?

**Comment by tedinski** [comment]
> Can you push this to a feature branch for now?

I thought this was a mergeable unit, or nearly so, so I requested the PR.

I could see merging this as-is, or possibly with a unit test for the irep generation added. We don't yet have many (any?) unit tests for cprover_bindings, so this might be the first, but we do actually want to get tests against this crate and have to start somewhere. :)

**Comment by celinval** [comment]
We already have unit tests in this crate. It would be great if we can add some food this code.

I would also suggest we tag the new functions with allow_dead with a link to the issue we're trying to solve.

**Comment by tautschnig** [comment]
I note that as of moments ago https://github.com/diffblue/cbmc/pull/6799 is merged, which means that function contracts now live in symbols of their own.

**Comment by monal** [comment]
Summary of changes: 

1. Add `contract` constructor to `Symbol` to create new symbols for function contracts.

2. Store the contract information (preconditions, postconditions, assigns, etc.) in the `value` field of the symbol (on the goto-program side). Hence, add `Contract(Contract)` to `ExprValue`. 
During `Irep` generation, the contract's preconditions and postconditions are added to the `type` field under `#spec_requires` and `#spec_ensures` respectively.

3. Add `LambdaExpression` and `Tuple` to `Expr`. They are mathematical expressions (Expressions that do not exist immediately in C) but adding them to `Expr` seems to be the best option right now. We can create a separate `MathematicalExpr` in the future, if needed. 
     * The clauses in `#spec_ensures(..)` and `#spec_requires(..)` are wrapped in a lambda expression (using the `as_lambda_expression` method) to make sure that the contract symbol is self-contained and has a copy of the functions’ parameters. (Refer: https://github.com/diffblue/cbmc/pull/6799).
    * The bound variables in a lambda expression are stored as a `Tuple`.

4. Add `MathematicalFunction` to `Type` to store the type signature of lambda expressions.

File: cprover_bindings/src/goto_program/expr.rs [hunk 1]
```diff
@@ -138,6 +143,10 @@ pub enum ExprValue {
     Symbol {
         identifier: InternedString,
     },
+    /// Tuple does not represent any expression in C.
```
[the remark this answers is not in the document]
Comment (danielsn) [reply cprover_bindings/src/goto_program/expr.rs:146]: It's basically an unnamed struct right?

File: cprover_bindings/src/goto_program/expr.rs [hunk 2]
```diff
@@ -411,6 +424,14 @@ impl Expr {
         expr!(Array { elems }, typ)
     }
 
+    pub fn lambda_expression(typ: Type, variables: Expr, body: Expr) -> Self {
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/expr.rs:427]: What is the type, and how do you make sure it matches the `variables` + `body`

File: cprover_bindings/src/goto_program/expr.rs [hunk 3]
```diff
@@ -411,6 +424,14 @@ impl Expr {
...
+        expr!(LambdaExpression { variables, body }, typ)
+    }
+
+    pub fn tuple_expr(typ: Type, operands: Vec<Expr>) -> Self {
+        expr!(Tuple { operands }, typ)
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/expr.rs:432]: What is the `typ` of a `tuple_expr`

File: cprover_bindings/src/goto_program/expr.rs [hunk 4]
```diff
@@ -1388,6 +1409,16 @@ impl Expr {
         expr!(StringConstant { s }, Type::c_char().array_of(s.len() + 1)).array_to_ptr()
     }
 }
+
+/// Wrap expression in a lambda
+impl Expr {
+    pub fn as_lambda_expression(&self, typ: Type, variables: &Vec<Expr>) -> Expr {
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/expr.rs:1415]: `variables: &[Expr]` is more general

File: cprover_bindings/src/goto_program/expr.rs [hunk 5]
```diff
@@ -1388,6 +1409,16 @@ impl Expr {
...
+        assert!(variables.iter().all(|x| x.is_symbol()), "Variables must be symbols");
+        let v = Expr::tuple_expr(Type::Empty, variables.to_vec());
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/expr.rs:1417]: Why type empty

File: cprover_bindings/src/goto_program/symbol.rs [hunk 6]
```diff
@@ -128,6 +129,28 @@ impl Symbol {
...
+        name: T,
+        base_name: U,
+        typ: Type,
+        contract: Contract,
+        loc: Location,
+    ) -> Symbol {
+        let name = name.into();
+        // Base name is the name of the function that the contract is written for.
+        let base_name: InternedString = base_name.into();
+        let pretty_name: InternedString = base_name;
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/symbol.rs:142]: Not the pretty name of the function the contract was written for?

File: cprover_bindings/src/goto_program/symtab_transformer/transformer.rs [hunk 7]
```diff
@@ -153,6 +156,19 @@ pub trait Transformer: Sized {
         transformed_typ.infinite_array_of()
     }
 
+    fn transform_type_mathematical_function(
+        &mut self,
+        domain: &Vec<Type>,
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/symtab_transformer/transformer.rs:161]: &[Type]

File: cprover_bindings/src/goto_program/symtab_transformer/transformer.rs [hunk 8]
```diff
@@ -153,6 +156,19 @@ pub trait Transformer: Sized {
...
+        codomain: &Type,
+    ) -> Type {
+        let transformed_domain: Vec<Type> = domain.iter().map(|x| self.transform_type(x)).collect();
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/symtab_transformer/transformer.rs:164]: Do you need the explicit type annotation here?

File: cprover_bindings/src/goto_program/typ.rs [hunk 9]
```diff
@@ -22,15 +22,24 @@ use std::fmt::Debug;
 #[derive(PartialEq, Debug, Clone)]
 pub enum Type {
     /// `typ x[size]`. E.g. `unsigned int x[3]`
-    Array { typ: Box<Type>, size: u64 },
+    Array {
+        typ: Box<Type>,
+        size: u64,
+    },
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/typ.rs:28]: Why the whitespace change?

File: cprover_bindings/src/goto_program/typ.rs [hunk 10]
```diff
@@ -39,37 +48,70 @@ pub enum Type {
...
-    IncompleteUnion { tag: InternedString },
+    IncompleteUnion {
+        tag: InternedString,
+    },
     /// CBMC specific. `typ x[__CPROVER_infinity()]`
-    InfiniteArray { typ: Box<Type> },
+    InfiniteArray {
+        typ: Box<Type>,
+    },
+    MathematicalFunction {
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/typ.rs:68]: Comment explaining what this is

File: cprover_bindings/src/irep/to_irep.rs [hunk 11]
```diff
@@ -480,10 +490,11 @@ impl ToIrep for SwitchCase {
 impl goto_program::Symbol {
     pub fn to_irep(&self, mm: &MachineModel) -> super::Symbol {
         super::Symbol {
-            typ: self.typ.to_irep(mm),
+            typ: self.typ.to_irep(mm).with_contract(&self.value, mm),
```
Comment (danielsn) [inline cprover_bindings/src/irep/to_irep.rs:493]: Does every `Symbol` have a contract?

File: cprover_bindings/src/goto_program/expr.rs (hunk 5 above)
Comment (zhassan-aws) [inline cprover_bindings/src/goto_program/expr.rs:1417]: Is `Type::Empty` intentional? If so, can you add a comment to explain why?

File: cprover_bindings/src/goto_program/typ.rs [hunk 12]
```diff
@@ -104,6 +146,12 @@ pub struct Parameter {
     base_name: Option<InternedString>,
 }
 
+/// Type for function contracts, loop contracts, etc.
+#[derive(Debug, Clone)]
+pub enum Contract {
```
Comment (zhassan-aws) [inline cprover_bindings/src/goto_program/typ.rs:151]: Does this need to be an enum?

File: cprover_bindings/src/goto_program/typ.rs [hunk 13]
```diff
@@ -104,6 +146,12 @@ pub struct Parameter {
...
+    FunctionContract { ensures: Vec<Expr>, requires: Vec<Expr>, variables: Vec<Expr> },
```
Comment (zhassan-aws) [inline cprover_bindings/src/goto_program/typ.rs:152]: Can you add a comment on what `variables` is?

File: cprover_bindings/src/goto_program/symbol.rs (hunk 6 above)
Comment (monal) [reply cprover_bindings/src/goto_program/symbol.rs:142]: No. Pretty name also contains the name of the function.
Reference: [c_typecheck_base.cpp](https://github.com/diffblue/cbmc/blob/b466503080639f98af3329e9e9bb05ac847a2fc3/src/ansi-c/c_typecheck_base.cpp) (lines 880-881)

File: cprover_bindings/src/goto_program/expr.rs (hunk 4 above)
Comment (monal) [reply cprover_bindings/src/goto_program/expr.rs]: Noted.

File: cprover_bindings/src/goto_program/expr.rs [hunk 14]
```diff
@@ -104,6 +104,11 @@ pub enum ExprValue {
     },
     /// `123`
     IntConstant(BigInt),
+    /// Lambda expressions do not exist immediately in C.
```
[the remark this answers is not in the document]
Comment (monal) [reply cprover_bindings/src/goto_program/expr.rs]: I've removed `LambdaExpression` from `Expr`. I'm generating the lambda expressions of the contract clauses at the `Irep` level now. That seems to be a better choice than exposing a generic `LambdaExpression` and `Tuple` at the`Expr` level.

File: cprover_bindings/src/goto_program/expr.rs [hunk 15]
```diff
@@ -104,6 +104,11 @@ pub enum ExprValue {
...
+    LambdaExpression {
+        variables: Expr,
```
[the remark this answers is not in the document]
Comment (monal) [reply cprover_bindings/src/goto_program/expr.rs]: It was meant to represent a `Tuple` of arguments, which in turn was an `Expr`. This code has now been removed.

File: cprover_bindings/src/goto_program/expr.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (monal) [reply cprover_bindings/src/goto_program/expr.rs]: Removed `Tuple` from `Expr`.

File: cprover_bindings/src/goto_program/typ.rs (hunk 9 above)
Comment (monal) [reply cprover_bindings/src/goto_program/typ.rs]: Fixed.

File: cprover_bindings/src/goto_program/expr.rs (hunk 5 above)
Comment (monal) [reply cprover_bindings/src/goto_program/expr.rs]: Removed `LambdaExpression` and this wrapper method from `Expr`.

File: cprover_bindings/src/goto_program/symtab_transformer/transformer.rs (hunk 8 above)
Comment (monal) [reply cprover_bindings/src/goto_program/symtab_transformer/transformer.rs]: Removed explicit type annotation after changing type of domain to &[Type].

File: cprover_bindings/src/goto_program/typ.rs (hunk 10 above)
Comment (monal) [reply cprover_bindings/src/goto_program/typ.rs]: Added comment.

File: cprover_bindings/src/goto_program/typ.rs (hunk 13 above)
Comment (monal) [reply cprover_bindings/src/goto_program/typ.rs]: Added comment. Moved contract to its own module.

File: cprover_bindings/src/goto_program/expr.rs (hunk 5 above)
Comment (monal) [reply cprover_bindings/src/goto_program/expr.rs]: Removed `LambdaExpression` from `Expr`. Adding tuple at the `Irep` level now.

File: cprover_bindings/src/goto_program/typ.rs (hunk 12 above)
Comment (monal) [reply cprover_bindings/src/goto_program/typ.rs]: Made it an enum to support different types of contracts. Like `LoopContract(...)` in the future. Is there an alternative way to do this?

**Comment by monal** [comment]
* Removed `LambdaExpression` and `Tuple` from `Expr`. Generating them at the Irep level now.
* Added a module for `Contract` and introduced `Spec` to represent contract clauses.
* Cleaned up whitespaces; added comments.

**Review by zhassan-aws — APPROVED** [review]
LGTM. Thanks!

File: cprover_bindings/src/goto_program/contract.rs [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        self.location = loc;
+        self
+    }
+}
+
+/// Constructor
+impl Spec {
+    pub fn new(temporary_symbols: Vec<Expr>, clause: Expr) -> Self {
+        assert!(temporary_symbols.iter().all(|x| x.is_symbol()), "Variables must be symbols");
+        Spec { temporary_symbols, clause, location: Location::none() }
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/contract.rs:54]: Do we want. `Location::none` by default?

File: cprover_bindings/src/goto_program/contract.rs [hunk 17]
```diff
@@ -0,0 +1,56 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
```
[the remark this answers is not in the document]
Comment (monal) [reply cprover_bindings/src/goto_program/contract.rs:3]: Added comment + CBMC documentation link.

File: cprover_bindings/src/goto_program/contract.rs (hunk 16 above)
Comment (monal) [reply cprover_bindings/src/goto_program/contract.rs]: Removed `Location::none`. Always setting the location in the constructor instead.

File: cprover_bindings/src/goto_program/symbol.rs [hunk 18]
```diff
@@ -128,6 +129,29 @@ impl Symbol {
...
+        name: T,
+        base_name: U,
+        typ: Type,
+        contract: Contract,
+        loc: Location,
+    ) -> Symbol {
+        let name = name.into();
+        // Both base name and pretty name contain the name of the function that the contract is written for.
+        let base_name: InternedString = base_name.into();
+        let pretty_name = base_name;
```
[the remark this answers is not in the document]
Comment (monal) [reply cprover_bindings/src/goto_program/symbol.rs:142]: On the CBMC side, the `pretty_name` is still the name of the function and not the name of the contract. Should I set it to the contract's name instead?

File: cprover_bindings/src/irep/to_irep.rs [hunk 19]
```diff
@@ -30,6 +30,23 @@ fn code_irep(kind: IrepId, ops: Vec<Irep>) -> Irep {
...
+fn lambda_irep(binding_variables: &[Expr], body: &Expr, mm: &MachineModel) -> Irep {
+    let typ = Type::MathematicalFunction {
+        domain: binding_variables.iter().map(|v| v.typ().clone()).collect::<Vec<_>>(),
+        codomain: Box::new(body.typ().clone()),
+    };
+    Irep {
+        id: IrepId::Lambda,
+        sub: vec![
+            tuple_irep(binding_variables, mm)
+                // For binding expressions (quantifies, let, lambda), the type of the `tuple_exprt` is set to just its ID in CBMC.
```
[the remark this answers is not in the document]
Comment (monal) [reply cprover_bindings/src/irep/to_irep.rs:43]: The `tuple_exprt` has a constructor with a `type` argument. So, in CBMC, we can construct tuples with a specific type (and not just the ID). The `binding_exprt` that sits on top of a `tuple_exprt` always sets the type to be the ID.

**Pull request merged**